libraries/lib-cloud-audiocom/UploadService.h

```diff
@@ -0,0 +1,98 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  UploadService.h
+
+  Dmitry Vedenko
+
+**********************************************************************/
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include <wx/string.h>
+
+namespace cloud::audiocom
+{
+class ServiceConfig;
+class OAuthService;
+
+//! Message that is sent when upload is finished.
+//! This message is sent from the network thread.
+struct CLOUD_AUDIOCOM_API UploadOperationCompleted final
+{
+   //! Result of the upload
+   enum class Result
+   {
+      //! Upload was successful
+      Success,
+      //! Upload was aborted by the user
+      Aborted,
+      //! Specified file is not found
+      FileNotFound,
+      //! Authorization is required
+      Unauthorized,
+      //! audio.com has failed to understand what Audacity wants
+      InvalidData,
+      //! Audacity has failed to understand audio.com response
+      UnexpetedResponse,
+      //! Upload failed for some other reason
+      UploadFailed
+   };
+   
+   //! Upload result
+   Result result;
+
+   //! Error message if result is not Success
+   std::string errorMessage;
+
+   //! URL to open in browser to finish up the anonymous upload
+   std::string finishUploadURL;
+   //! "Slug" to be used for sharable URL construction
+   std::string audioSlug;
+};
+
+//! Class used to track the upload operation
+class CLOUD_AUDIOCOM_API UploadPromise
+{
+public:
+   virtual ~UploadPromise();
+
+   //! Returns true if the upload is finished
+   virtual bool IsCompleted() = 0;
+   //! Abort the upload, if running
+   virtual void Abort() = 0;
+   //! Abort the upload, if running,
+   //! notify audio.com that the uploaded file is no longer needed otherwise 
+   virtual void DiscardResult() = 0;
```

Comment:
What should happen in case this "promise" (or future) is destroyed before completion, without any calls to `Abort` or `DiscardResult` ?  It could be that `UploadOperation` still occupies some resources unnecessarily because `mAborted` isn't true.

Should `UploadPromise` call `Abort()` in its detructor if not completed?  What are you crazy?  That would be a pure virtual function call!  The base class can't do that...